sympy/core/mod.py

```diff
@@ -107,6 +107,17 @@ def doit(p, q):
             elif (qinner*(q + qinner)).is_nonpositive:
                 # |qinner| < |q| and have different sign
                 return p
+        elif any(isinstance(arg, cls) for arg in p.args):
+            #separating into modulus and non modulus
+            both_l = non_mod_l, mod_l = [], []
+            for arg in p.args:
+                both_l[isinstance(arg,cls)].append(arg)
+            #if q same for all
+            if all(inner.args[1] == q for inner in mod_l) and \
+                    all(cls(z,q) == z for z in non_mod_l):
```

Comment:
I don't think this last check is necessary. Mod(Mod(A, c) + B, c) == Mod(A + B, c) unconditionally.